ws-server/src/server.rs

```diff
@@ -73,43 +73,101 @@ impl Server {
 
 	/// Start responding to connections requests. This will block current thread until the server is stopped.
 	pub async fn start(self) {
-		let mut incoming = TcpListenerStream::new(self.listener);
 		let methods = self.methods;
-		let conn_counter = Arc::new(());
 		let mut id = 0;
 
-		while let Some(socket) = incoming.next().await {
-			if let Ok(socket) = socket {
-				if let Err(e) = socket.set_nodelay(true) {
-					log::error!("Could not set NODELAY on socket: {:?}", e);
-					continue;
-				}
+		let mut driver = ConnDriver::new(self.listener);
+
+		loop {
+			match Pin::new(&mut driver).await {
+				Ok((socket, _addr)) => {
+					if let Err(e) = socket.set_nodelay(true) {
+						log::error!("Could not set NODELAY on socket: {:?}", e);
+						continue;
+					}
+
+					if driver.count() >= self.cfg.max_connections as usize {
+						log::warn!("Too many connections. Try again in a while");
+						continue;
+					}
+
+					let methods = &methods;
+					let cfg = &self.cfg;
 
-				if Arc::strong_count(&conn_counter) > self.cfg.max_connections as usize {
-					log::warn!("Too many connections. Try again in a while");
-					continue;
+					driver.add(Box::pin(background_task(socket, id, methods, cfg)));
+
+					id += 1;
+				}
+				Err(err) => {
+					log::error!("Error while awaiting a new connection: {:?}", err);
 				}
-				let methods = methods.clone();
-				let counter = conn_counter.clone();
-				let cfg = self.cfg.clone();
+			}
+		}
+	}
+}
+
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// This is a glorified select `Future` that will attempt to drive all
+/// connection futures `F` to completion on each `poll`, while also
+/// handling incoming connections.
+struct ConnDriver<F> {
+	listener: TcpListener,
+	connections: Vec<F>,
+}
 
-				tokio::spawn(async move {
-					let r = background_task(socket, id, methods, cfg).await;
-					drop(counter);
-					r
-				});
+impl<F> ConnDriver<F>
+where
+	F: Future + Unpin,
+{
+	fn new(listener: TcpListener) -> Self {
+		ConnDriver { listener, connections: Vec::new() }
+	}
 
-				id += 1;
+	fn count(&self) -> usize {
+		self.connections.len()
+	}
+
+	fn add(&mut self, conn: F) {
+		self.connections.push(conn);
+	}
+}
+
+impl<F> Future for ConnDriver<F>
+where
+	F: Future + Unpin,
+{
+	type Output = std::io::Result<(TcpStream, SocketAddr)>;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		let this = Pin::into_inner(self);
+
+		let mut i = 0;
+
+		while i < this.connections.len() {
+			if this.connections[i].poll_unpin(cx).is_ready() {
```

Comment:
So when we get `Poll::Ready` here it means the background task has finished/exited?

Reply:
Correct.